### PR TITLE
DAOS-11403 common: Add unit testing and fix bugs

### PR DESCRIPTION
**Pull request by jolivier23**
Rearrange the unit test to make it a little simpler to add new tests Add some unit testing for checkpointing
Fix a bug where we could overflow the chkpt_data buffer Add umem_touch call in replay_cb

Required-githooks: true

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/common/dav/wal_tx.c [hunk 1]
```diff
@@ -482,6 +485,7 @@ dav_wal_replay_cb(uint64_t tx_id, struct umem_action *act, void *base)
 			act->ac_op_bits.addr / PAGESIZE, act->ac_op_bits.addr % PAGESIZE,
 			act->ac_op_bits.pos, act->ac_op_bits.num);
 		off = act->ac_op_bits.addr;
+		size = sizeof(uint64_t);
```
Comment (NiuYawei) [inline src/common/dav/wal_tx.c:488]: This set/clear bits only supports sizeof(uint64_t)? It's not from this PR, but I don't understand the "ASSERT_rt((pos >=0 && (pos +,num) <= 64) below.

File: src/common/dav/dav_iface.c [hunk 2]
```diff
@@ -156,8 +156,7 @@ dav_obj_open_internal(int fd, int flags, size_t sz, const char *path, struct ume
 
 		/* REVISIT: checkpoint case */
 		D_ASSERT(store != NULL);
-		rc = hdl->do_store->stor_ops->so_wal_replay(hdl->do_store, dav_wal_replay_cb,
-							    hdl->do_base);
+		rc = hdl->do_store->stor_ops->so_wal_replay(hdl->do_store, dav_wal_replay_cb, hdl);
```
Comment (NiuYawei) [inline src/common/dav/dav_iface.c:159]: Load meta data to page cache before replay?

File: src/common/dav/wal_tx.c (hunk 1 above)
Comment (NiuYawei) [reply src/common/dav/wal_tx.c:488]: Oh, never mind.

File: src/include/daos_srv/bio.h [hunk 3]
```diff
@@ -966,7 +966,8 @@ struct bio_io_context *bio_mc2ioc(struct bio_meta_context *mc, enum smd_dev_type
  *
  * \return			Zero on success, negative value on error
  */
-int bio_wal_reserve(struct bio_meta_context *mc, uint64_t *tx_id);
+int
+     bio_wal_reserve(struct bio_meta_context *mc, uint64_t *tx_id);
```
Comment (NiuYawei) [inline src/include/daos_srv/bio.h:970]: Is this an unintentional change caused by commit script?

File: src/common/dav/dav_iface.c [hunk 4]
```diff
@@ -22,6 +22,52 @@
...
+		umemstor_sgl_set(&sgl, &iov, start, read_size);
+		rc = store->stor_ops->so_read(store, &iod, &sgl);
+		if (rc)
+			break;
+		off += read_size;
+		start += read_size;
+		remain_size -= read_size;
+	}
+	return rc;
+}
```
Comment (NiuYawei) [inline src/common/dav/dav_iface.c:69]: I think Sherin will update this to use optimized read size and removing the global vos_pmemobj_lock for BMEM and ADMEM, will the update being incorporated?

File: src/common/dav/dav_iface.c [hunk 5]
```diff
@@ -22,6 +22,52 @@
 #define	DAV_HEAP_INIT	0x1
 #define MEGABYTE	((uintptr_t)1 << 20)
 
+#define META_READ_BATCH_SIZE (16 * 1024 * 1024)
```
Comment (sherintg) [inline src/common/dav/dav_iface.c:25]: This size can be changed to 1M as it gives the most optimal read throughput.

File: src/common/dav/dav_iface.c (hunk 4 above)
Comment (sherintg) [reply src/common/dav/dav_iface.c:69]: I think we can remove the use of vos_pmemobj_lock. I checked the pmdk code that is built with daos and it uses a global mutex lock (pools_mutex) to serialize calls to create/open/close calls.

**Review by sherintg — CHANGES_REQUESTED** [review]

File: src/include/daos_srv/bio.h (hunk 3 above)
Comment (jolivier23) [reply src/include/daos_srv/bio.h:970]: yeah, I don't know why clang-format does this.  Probably a bug.  I can try to revert

**Review by NiuYawei — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Review by sherintg — APPROVED** [review]
I will make the changes related to removing the vos_pmemobj_lock. Rest of the changes looks fine.

**Pull request merged**